### PR TITLE
Add Echo task

### PR DESCRIPTION
**Pull request by pingsutw**
## Tracking issue
https://github.com/flyteorg/flyte/issues/3533

## Why are the changes needed?
Currently, the conditional in Flyte does not support skipping one of the branch without any task execution. This means that we will need to use a noop task to skip the branch. This will lead to a waste of time during execution.

## What changes were proposed in this pull request?
Add an echo task inside flytekit, and propeller will use [echo](https://github.com/flyteorg/flyte/blob/master/flyteplugins/go/tasks/plugins/testing/echo.go) plugin to run it without launching a pod.

`echo` task just simply copy the inputs to outputs and return to the users

## How was this patch tested?
```python
import typing
from click.testing import CliRunner

from flytekit import task, workflow, ImageSpec, conditional
from flytekit.clis.sdk_in_container import pyflyte
from flytekit.core.task import Echo

new_flytekit = "git+https://github.com/flyteorg/flytekit.git@5d6e9a0d2882d7484db682e065dfa269d8a02964"

custom_image = ImageSpec(
    registry="ghcr.io/flyteorg",
    packages=[new_flytekit],
    apt_packages=["git"],
)


echo = Echo(name="echo", inputs={"a": float})


@task(container_image=custom_image)
def calculate_circle_circumference(radius: float) -> typing.Optional[float]:
    return 2 * 3.14 * radius  # Task to calculate the circumference of a circle


@task(container_image=custom_image)
def calculate_circle_area(radius: float) -> int:
    return 3.14 * radius * radius  # Task to calculate the area of a circle


@workflow
def wf(radius: float) -> typing.Optional[float]:
    return (
        conditional("shape_properties_with_multiple_branches")
        .if_((radius >= 0.1) & (radius < 1.0), last_case=False)
        .then(calculate_circle_circumference(radius=radius))
        .else_()
        .then(echo(a=radius))
    )


if __name__ == '__main__':
    runner = CliRunner()
    result = runner.invoke(
        pyflyte.main,
        [
            "register",
            "/Users/kevin/git/flytekit/flyte-example/condition.py",
        ]
    )
    print(result.stdout)
```

### Setup process
Enable the `echo` plugin in the flytepropeller.
```yaml
tasks:
  task-plugins:
    enabled-plugins:
      - container
      - sidecar
      - K8S-ARRAY
      - agent-service
      - echo                                                                                                                                                                                                                       
      # - ray
    default-for-task-types:
      - container: container
      - container_array: K8S-ARRAY
```

### Screenshots
![image](https://github.com/user-attachments/assets/51ffbc21-138d-44fe-ac8f-be681e259e3a)

## Check all the applicable boxes <!-- Follow the above conventions to check the box -->

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
NA

## Docs link
NA

### CODE REVIEW

**Comment by pingsutw** [comment]
cc @yubofredwang

File: flytekit/core/task.py [hunk 1]
```diff
@@ -416,3 +416,30 @@ def wrapper(fn) -> ReferenceTask:
...
+    A task that simply echoes the inputs back to the user.
+    The task's inputs and outputs interface are the same.
+    FlytePropeller won't create a pod for this task, it will simply pass the inputs to the outputs.
+    https://github.com/flyteorg/flyte/blob/master/flyteplugins/go/tasks/plugins/testing/echo.go
+    """
+
+    _TASK_TYPE = "echo"
+
+    def __init__(self, name: str, inputs: Optional[Dict[str, Type]] = None, **kwargs):
+        outputs = dict(zip(list(output_name_generator(len(inputs.values()))), inputs.values())) if inputs else None
```
Comment (thomasjpfan) [inline flytekit/core/task.py:432]: Nit:

```suggestion
        outputs = dict(zip(output_name_generator(len(inputs)), inputs.values())) if inputs else None
```

File: flytekit/core/task.py [hunk 2]
```diff
@@ -416,3 +416,30 @@ def wrapper(fn) -> ReferenceTask:
...
 
     return wrapper
+
+
+class Echo(PythonTask):
+    """
+    A task that simply echoes the inputs back to the user.
+    The task's inputs and outputs interface are the same.
+    FlytePropeller won't create a pod for this task, it will simply pass the inputs to the outputs.
+    https://github.com/flyteorg/flyte/blob/master/flyteplugins/go/tasks/plugins/testing/echo.go
```
Comment (thomasjpfan) [inline flytekit/core/task.py:426]: Can we include docs for `name`, `inputs`, and `kwargs`?

File: tests/flytekit/unit/core/test_conditions.py [hunk 3]
```diff
@@ -495,3 +496,41 @@ def multiplier_2(my_input: float) -> float:
...
+    assert wf1(radius=1.8) == 1.8
+
+    echo2 = Echo(name="echo", inputs={"a": float, "b": float})
+
+    @task()
+    def t2(radius: float) -> (float, float):
+        return 2 * 3.14 * radius, 2 * 3.14 * radius
+
+    @workflow
+    def wf2(radius1: float, radius2: float) -> (float, float):
```
Comment (thomasjpfan) [inline tests/flytekit/unit/core/test_conditions.py:527]: I do not think a typing with a tuple like this is support in Python: https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#useful-built-in-types

```suggestion
    def wf2(radius1: float, radius2: float) -> typing.Tuple[float, float]:
```

**Review by thomasjpfan — COMMENTED** [review]
When I run `pyflyte run --remote`:

```
import typing

from flytekit import ImageSpec, conditional, task, workflow
from flytekit.core.task import Echo

new_flytekit = "git+https://github.com/flyteorg/flytekit.git@68c6cf91891923c4cc898b2a493d42927035ce32"

custom_image = ImageSpec(
    registry="localhost:30000",
    packages=[new_flytekit],
    apt_packages=["git"],
)


echo1 = Echo(name="echo", inputs={"a": float})


@task(container_image=custom_image)
def t1(radius: float) -> typing.Optional[float]:
    return 2 * 3.14 * radius


@workflow
def wf1(radius: float) -> typing.Optional[float]:
    return (
        conditional("shape_properties_with_multiple_branches")
        .if_((radius >= 0.1) & (radius < 1.0))
        .then(t1(radius=radius))
        .else_()
        .then(echo1(a=radius))
    )
```

I am getting this error:

```
Request rejected by the API, due to Invalid input.
RPC Failed, with Status: StatusCode.INVALID_ARGUMENT
        details: failed to compile workflow for [resource_type:WORKFLOW  project:"flytesnacks"  domain:"development"  name:"tjpf_noop.wf1"  version:"bonkFHVywVrbD1ngx7SsDQ"] with err failed to compile workflow with err Collected Errors: 5
        Error 0: Code: MismatchingTypes, Node Id: n0-n1, Description: Variable [o0] (type [union_type:{variants:{simple:FLOAT  structure:{tag:"float"}}  variants:{simple:NONE  structure:{tag:"none"}}}]) doesn't match expected type [simple:FLOAT].
        Error 1: Code: ParameterNotBound, Node Id: end-node, Description: Parameter not bound [o0].
        Error 2: Code: UnreachableNodes, Node Id: start-node, Description: The Workflow contain unreachable nodes [n0].
        Error 3: Code: ValueRequired, Node Id: n0, Description: Value required [.radius].
        Error 4: Code: VariableNameNotFound, Node Id: n0, Description: Variable [o0] not found on node [n0].
```

**Comment by pingsutw** [comment]
sorry, my bad.
```python
echo1 = Echo(name="echo", inputs={"a": float})
```

it should be
```python
echo1 = Echo(name="echo", inputs={"a": typing.Optional[float]})
```

File: tests/flytekit/unit/core/test_conditions.py (hunk 3 above)
Comment (pingsutw) [reply tests/flytekit/unit/core/test_conditions.py:527]: Updated it, thanks. I saw we use `-> (...)` in many places in flytekit.

File: flytekit/core/task.py (hunk 2 above)
Comment (pingsutw) [reply flytekit/core/task.py]: updated

File: flytekit/core/task.py (hunk 1 above)
Comment (pingsutw) [reply flytekit/core/task.py]: done

**Comment by thomasjpfan** [comment]
When I run:

```python
import typing

from flytekit import ImageSpec, conditional, task, workflow
from flytekit.core.task import Echo

new_flytekit = "git+https://github.com/flyteorg/flytekit.git@2349ceeaf74fe6449b66eee54be5667aec0b45ab"

custom_image = ImageSpec(
    registry="localhost:30000",
    packages=[new_flytekit],
    apt_packages=["git"],
    builder="default",
)


echo1 = Echo(name="echo", inputs={"a": typing.Optional[float]})


@task(container_image=custom_image)
def t1(radius: float) -> typing.Optional[float]:
    return 2 * 3.14 * radius


@workflow
def wf1(radius: float) -> typing.Optional[float]:
    return (
        conditional("shape_properties_with_multiple_branches")
        .if_((radius >= 0.1) & (radius < 1.0))
        .then(t1(radius=radius))
        .else_()
        .then(echo1(a=radius))
    )
```

I get this error:

```
Workflow[flytesnacks:development:main.wf1] failed. RuntimeExecutionError: max number of system retry attempts [11/10] exhausted. Last known status message: failed at Node[n0-n1]. RuntimeExecutionError: failed during plugin execution, caused by: failed to execute handle for plugin [container]: [BadTaskSpecification] invalid TaskSpecification, unable to determine Pod configuration
```

Screenshot 2024-08-16 at 11 36 39 AM

**Comment by pingsutw** [comment]
It's not enabled by default in the backend. Need to update the config. just updated the PR description, sorry.
```
tasks:
  task-plugins:
    enabled-plugins:
      - container
      - sidecar
      - K8S-ARRAY
      - agent-service
      - echo                                                                                                                                                                                                                       
      # - ray
    default-for-task-types:
      - container: container
      - container_array: K8S-ARRAY
```

File: flytekit/core/task.py [hunk 4]
```diff
@@ -416,3 +416,35 @@ def wrapper(fn) -> ReferenceTask:
...
+
+class Echo(PythonTask):
+    _TASK_TYPE = "echo"
+
+    def __init__(self, name: str, inputs: Optional[Dict[str, Type]] = None, **kwargs):
+        """
+        A task that simply echoes the inputs back to the user.
+        The task's inputs and outputs interface are the same.
+        FlytePropeller won't create a pod for this task, it will simply pass the inputs to the outputs.
+        https://github.com/flyteorg/flyte/blob/master/flyteplugins/go/tasks/plugins/testing/echo.go
```
Comment (thomasjpfan) [inline flytekit/core/task.py:429]: As a follow up, I think this should link out to a docs around "How to enable echo tasks". This way a user can look at the docstring and know to contact their platform engineer to enable the feature.

Comment (pingsutw) [reply flytekit/core/task.py:429]: Thanks! good call.
- [PR](https://github.com/flyteorg/flytesnacks/pull/1724) for 1
- [PR](https://github.com/flyteorg/flyte/pull/5679) for 2

Comment (pingsutw) [reply flytekit/core/task.py:429]: I will create another follow-up PR to address this bad error. 

image

Comment (pingsutw) [reply flytekit/core/task.py:431]: I just updated the docstring in this PR too

**Review by wild-endeavor — APPROVED** [review]
In the future we can improve on this a tiny bit (though hard to justify I think), but we can create a hardcoded no-op marker object, and whenever the conditional sees a no-op marker, it just automatically creates this echo task for the user.

**Pull request merged**